### PR TITLE
LPS-145947 Creates a new *not public* editor component called BaseEditor and use it

### PR DESCRIPTION
**Pull request by diegonvs**
This PR is an alternative version of https://github.com/liferay-frontend/liferay-portal/pull/1871 where We don't need to deprecate Editor component anymore and We can add our shared logic/code on BaseEditor component as It was proposed on: https://github.com/liferay-frontend/liferay-portal/pull/1871/commits/f841413315a8260e39441799ef8bbf6722bef054

Screen Shot 2022-02-04 at 17 38 32

### CODE REVIEW

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/insertbutton/plugin.js [hunk 1]
```diff
@@ -166,13 +166,13 @@
 
 			tableToolbar.hide();
 
-			setTimeout(() => {
+			requestAnimationFrame(() => {
```
Comment (julien) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/insertbutton/plugin.js:169]: always better than `setTimeout` :+1:

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BaseEditor.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import PropTypes from 'prop-types';
+import React, {forwardRef, useCallback, useEffect, useRef} from 'react';
+
+import {Editor} from './Editor';
```
Comment (markocikos) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BaseEditor.js:18]: Keeping a dependency to `Editor` is an interesting architectural proposal, but I don't think is the best way to go:
- In my view, there is no difference between `getHTML`-related code and `beforeScreenFlip` code. Both are internal DXP-related tweaks / quirks, that should be kept under the hood. Both are common to **all** variants of CKeditor usages. There is no need for two layers of base code.
- We want all variants of CKeditor to have all common DXP-specific code. The way would be in this PR, `Editor` would be missing `getHTML` stuff.
- We want to support `ClassicEditor` as a thin wrapper to CKeditor in DXP. We don't want an odd alternative with only some DXP fixes. As such, we want to deprecate `Editor`, and in a future iteration, get rid of it. If we keep dependency like in this PR, we cannot deprecate/remove it.

What we should do here is:
- `BaseEditor` should import `ckeditor4-react`.
- `beforeScreenFlip` should be in `BaseEditor`.
- `Editor` should say exactly as it is, so we don't break backwards compatibility.
- We should deprecate `Editor`.

**Review by markocikos — CHANGES_REQUESTED** [review]
@diegonvs see comment inline

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/insertbutton/plugin.js (hunk 1 above)
Comment (diegonvs) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/insertbutton/plugin.js:169]: 🫂

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BaseEditor.js (hunk 2 above)
Comment (dsanz) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BaseEditor.js]: I like the idea of abstracting out common stuff in a `BaseEditor` and do not export it in the component `index.js`.

I'm a little bit in doubt about keeping `Editor.js` as part of all this new architecture. On one hand, "Editor" is a name that does not suggest what we wanted ("do not use it as it's an internal thing"), so it's probably better to deprecate it and add a comment in the file itself stating the intent. It's not a big deal if we write `BaseEditor` with no dependency on `Editor`.

But on the other, a proper deprecation requires usage analysis and determining whether existing uses can be replaced by some "other" thing. And that's where my doubts arise. What that other thing would look like? Some new variant on top of `BaseEditor`? or maybe just a plain, bare `ckeditor4-react`, with no support for DXP machinery (is it even a good idea)? The alleged purpose of this task is to offer a unified, consistent editor DXP API covering all the cases we know of. Could it be a potential outcome after the analysis is that we can't "deprecate" what `Editor` provides? (I'm hoping it's not)

**Comment by diegonvs** [comment]
Screen Shot 2022-02-07 at 12 06 14

The status of https://github.com/liferay-frontend/liferay-portal/pull/1891/commits/e39c417ee614038c63c27806443976ab8a2a2280

**Comment by diegonvs** [comment]
ci:test:relevant

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BaseEditor.js (hunk 2 above)
Comment (markocikos) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BaseEditor.js]: >But on the other, a proper deprecation requires usage analysis and determining whether existing uses can be replaced by some "other" thing. And that's where my doubts arise. What that other thing would look like? Some new variant on top of BaseEditor? or maybe just a plain, bare ckeditor4-react, with no support for DXP machinery (is it even a good idea)? The alleged purpose of this task is to offer a unified, consistent editor DXP API covering all the cases we know of. Could it be a potential outcome after the analysis is that we can't "deprecate" what Editor provides? (I'm hoping it's not)

Yeah, valid concerns! I'd say that it's almost certain that `ClassicEditor` will be perfectly fine. There is a tiny chance that is not, if usages do something related to DXP quirks in common code. We can analyze, test and migrate usages in a separate task. We should do that asap after this PR is merged.

**Comment by markocikos** [comment]
Awesome graphs, btw! :heart_eyes:

**Comment by diegonvs** [comment]
ci:test:relevant

**Comment by diegonvs** [comment]
ci:stop

**Comment by diegonvs** [comment]
ci:test:relevant

**Comment by diegonvs** [comment]
https://github.com/liferay-frontend/liferay-portal/pull/1891/commits/5ae5bfb32e36510d52db6030b075ee7260f5807b fixes the CI errors, re-running relevant suite :)

**Comment by markocikos** [comment]
ci:forward

**Pull request closed without merging**